File: protocol-generator/haskell/src/EntityCore/Identity.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Identity (L1) — a peer's keypair and the entities derived from it
-- (§1.5, §3.5, §7.3, §7.4). The peer's identity is a 32-byte Ed25519 seed;
-- everything else derives:
--
-- @
--   public_key    = Ed25519 pub of seed                       (32 bytes)
--   peer_id       = Base58(varint(1) ‖ varint(0) ‖ public_key)  (§1.5 identity-multihash)
--   peer entity   = system/peer { public_key, key_type }        (§3.5; v7.65 — NO
--                   peer_id in the hashable basis)
--   identity_hash = content_hash(peer entity)
-- @
--
-- Signing is over the full 33-byte content_hash (format byte + digest, §7.3), so
-- a signature is bound to the hash format.
--
-- §1.5/§7.4 reconciliation (the v7.74 E1 erratum, A-HS-NNN corroboration):
-- Ed25519 peer_id is identity-multihash form (key_type 0x01, hash_type 0x00,
-- digest = the raw public key). §7.4's older "NORMATIVE" SHA-256-form pseudocode
-- is superseded by the §1.5 table in v7.74; we follow §1.5 (derivePeerId).
module EntityCore.Identity
  ( Identity (..)
  , identityOfSeed
  , peerEntityOfPubkey
  , peerIdOfPubkey
  , signEntity
  , verifySignature
  , ed25519VerifyRaw
  ) where

import Data.ByteString (ByteString)
import Data.Text (Text)
import qualified Data.Text.Encoding as TE

import EntityCore.Codec.Value (Value (..))
import EntityCore.Model (Entity (..), bytesField, makeEntity)
import EntityCore.PeerId (derivePeerId, formatPeerId)
import qualified EntityCore.Signature as Sig

-- | A peer's local identity. All derived fields are precomputed at construction.
data Identity = Identity
  { idSeed :: ByteString -- ^ 32-byte Ed25519 seed
  , idPublicKey :: ByteString -- ^ 32 bytes
  , idPeerId :: Text -- ^ Base58 (ASCII)
  , idPeerEntity :: Entity
  , idIdentityHash :: ByteString -- ^ content_hash(peer entity), 33 bytes
  }

-- | The @system/peer@ entity for a public key (§3.5; v7.65 — no peer_id field in
-- the hashable basis).
peerEntityOfPubkey :: ByteString -> Entity
peerEntityOfPubkey publicKey =
  makeEntity
    "system/peer"
    (VMap [(VText "public_key", VBytes publicKey), (VText "key_type", VText "ed25519")])

-- | The Ed25519 canonical peer_id (identity-multihash, §1.5): key_type 0x01,
-- hash_type 0x00, digest = the raw public key.
peerIdOfPubkey :: ByteString -> Text
peerIdOfPubkey publicKey = TE.decodeUtf8 (formatPeerId (derivePeerId 0x01 publicKey))

-- | Build the local identity from a 32-byte seed. Errors only on a malformed
-- seed length (caught at host startup, not on the wire).
identityOfSeed :: ByteString -> Either String Identity
identityOfSeed seed = do
  publicKey <- either (Left . show) Right (Sig.ed25519PubkeyFromSeed seed)
  let peerEntity = peerEntityOfPubkey publicKey
  Right
    Identity
      { idSeed = seed
      , idPublicKey = publicKey
      , idPeerId = peerIdOfPubkey publicKey
      , idPeerEntity = peerEntity
      , idIdentityHash = entHash peerEntity
      }

-- | Sign an entity's content_hash; produce the @system/signature@ entity (§3.5):
-- target = signed entity hash, signer = our identity hash.
signEntity :: Identity -> Entity -> Entity
signEntity ident target =
  let !sigBytes = either (const "") id (Sig.ed25519Sign (idSeed ident) (entHash target))
   in makeEntity
        "system/signature"
        ( VMap
            [ (VText "target", VBytes (entHash target))
            , (VText "signer", VBytes (idIdentityHash ident))
            , (VText "algorithm", VText "ed25519")
            , (VText "signature", VBytes sigBytes)
            ]
        )

-- | Verify a @system/signature@ entity against the signer's @system/peer@ entity.
-- Reads public_key from the peer entity; the §5.2 signer-hash check is the
-- caller's responsibility.
verifySignature :: Entity -> Entity -> Bool
verifySignature signature signerPeer =
  case (bytesField signature "target", bytesField signature "signature", bytesField signerPeer "public_key") of
    (Just target, Just sigBytes, Just pub) ->
      either (const False) id (Sig.ed25519Verify pub target sigBytes)
    _ -> False

-- | Verify a raw Ed25519 signature over @msg@ under @pubkey@ (the §4.6
-- proof-of-possession check, where there is no @system/signature@ entity yet).
ed25519VerifyRaw :: ByteString -> ByteString -> ByteString -> Either String Bool
ed25519VerifyRaw pubkey msg sigBytes =
  either (Left . show) Right (Sig.ed25519Verify pubkey msg sigBytes)

File: protocol-generator/python/src/entity_core/peer/identity.py
"""Identity (L1): a peer's keypair + derived entities (§1.5, §3.5, §7.3).

A peer's identity is an Ed25519 seed; everything derives:

    public_key    = Ed25519 pub of seed                 (32 bytes)
    peer_id       = §1.5 canonical form (identity-multihash; hash_type=0x00)
    peer_entity   = system/peer {public_key, key_type}  (§3.5; v7.65 — NO
                    peer_id field in the hashable basis)
    identity_hash = content_hash(peer_entity)

Signing is over the full 33-byte content_hash (format byte + digest, §7.3) so a
signature is bound to the hash format.

§1.5 canonical peer-id derivation (A-SW-008 erratum; NOT the stale §7.4
SHA256(pubkey) skeleton): a key <= 32 bytes is identity-multihash form
(hash_type=0x00, digest = the raw key); a larger key is SHA-256-form
(hash_type=0x01, digest = SHA-256(key)).  Ed25519 (32 B) -> (0x01, 0x00, pub).
"""

from __future__ import annotations

from dataclasses import dataclass

from ..peer_id import format_peer_id, parse_peer_id
from ..signature import (
    ED25519_SEED_SIZE,
    ed25519_public_key,
    sign_ed25519,
    verify_ed25519,
)
from .model import Entity

#: Key types (§1.5).
KEY_TYPE_ED25519 = 0x01
KEY_TYPE_ED448 = 0x02
#: Hash types (§1.5 canonical-form table).
HASH_TYPE_IDENTITY = 0x00  # identity-multihash: digest IS the raw key
HASH_TYPE_SHA256 = 0x01  # SHA-256-form: digest = SHA-256(key)


def peer_id_of_public_key(public_key: bytes, key_type: int = KEY_TYPE_ED25519) -> str:
    """Derive the canonical Base58 peer-id for a raw public key (§1.5)."""
    if len(public_key) <= 32:
        return format_peer_id(key_type, HASH_TYPE_IDENTITY, public_key)
    import hashlib

    return format_peer_id(key_type, HASH_TYPE_SHA256, hashlib.sha256(public_key).digest())


def peer_entity_of_public_key(public_key: bytes) -> Entity:
    """Build the system/peer entity for a raw public key (v7.65: NO peer_id in
    the hashable basis — only {public_key, key_type})."""
    return Entity.make(
        "system/peer",
        {"public_key": bytes(public_key), "key_type": "ed25519"},
    )


@dataclass(frozen=True, slots=True)
class Identity:
    """A peer's keypair plus its derived peer entity and identity hash."""

    seed: bytes
    public_key: bytes
    peer_id: str
    peer_entity: Entity

    @staticmethod
    def of_seed(seed: bytes) -> "Identity":
        """Construct an identity from a 32-byte Ed25519 seed."""
        if len(seed) != ED25519_SEED_SIZE:
            raise ValueError(f"ed25519 seed must be {ED25519_SEED_SIZE} bytes")
        pub = ed25519_public_key(seed)
        return Identity(
            seed=bytes(seed),
            public_key=pub,
            peer_id=peer_id_of_public_key(pub),
            peer_entity=peer_entity_of_public_key(pub),
        )

    @property
    def identity_hash(self) -> bytes:
        """The content_hash of the peer entity (§3.5)."""
        return self.peer_entity.hash

    def sign_entity(self, target: Entity) -> Entity:
        """Sign ``target``'s content_hash, producing the system/signature entity
        (§3.5): target = signed hash, signer = our identity hash, signature =
        Ed25519 over the 33-byte content_hash."""
        sig = sign_ed25519(self.seed, target.hash)
        return Entity.make(
            "system/signature",
            {
                "target": bytes(target.hash),
                "signer": bytes(self.identity_hash),
                "algorithm": "ed25519",
                "signature": sig,
            },
        )


def verify_signature(signature: Entity, signer_peer: Entity) -> bool:
    """Verify a system/signature entity against the signer's system/peer entity.

    Reads public_key from the peer entity; the §5.2 signer-hash binding check is
    the caller's responsibility.
    """
    target = signature.bytes_("target")
    sig = signature.bytes_("signature")
    pub = signer_peer.bytes_("public_key")
    if target is None or sig is None or pub is None or len(pub) != 32:
        return False
    return verify_ed25519(pub, sig, target)


__all__ = [
    "Identity",
    "peer_id_of_public_key",
    "peer_entity_of_public_key",
    "verify_signature",
    "parse_peer_id",
    "KEY_TYPE_ED25519",
    "KEY_TYPE_ED448",
]
